### DISCUSSION TITLE
AvaloniaView in an iOS Extension

### DISCUSSION THREAD

**Discussion by tkefauver** [AvaloniaUI/Avalonia#16238] [Q&A]
I'm making an iOS custom keyboard and can't get an `AvaloniaView` to load properly into the `UIInputViewController.View` from `viewDidLoad()`. It crashes unless I create the `AvaloniaView` in a `Dispatcher.UIThread` frame but doing `UIInputViewController.View.AddSubView` inside `Dispatcher.UIThread` appears to have no affect when showing the keyboard.

Also should mention I implemented `IAvaloniaViewController` and call `AvaloniaView.InitWithController` but that doesn't change anything either.

Does anyone have any suggestions?

**Comment by tkefauver** [comment]
Maybe actual code might help. 

This crashes the extension (log says `Connection to plugin interrupted while in use`):
```
public class KeyboardViewController : UIInputViewController {
...
     public override void ViewDidLoad() {
         base.ViewDidLoad();
         InputView.AddSubview(new AvaloniaView());
     }
}
```

This has no affect:
```
public class KeyboardViewController : UIInputViewController {
...
     public override void ViewDidLoad() {
         base.ViewDidLoad();
         Dispatcher.UIThread.Post(()=>{
                InputView.AddSubview(new AvaloniaView());
         });
     }
}
```
Is there someway to associate Avalonia w/ an existing ViewController? Keyboard extensions aren't normal apps (no `AppDelegate` or `Application`) so I think I maybe out of luck here.

**Reply by maxkatz6** [reply]
How do you setup/initialize Avalonia in this project?
If you have a custom AvaloniaAppDelegate, you still need create and build AppBuilder. Possibly re-implementing parts of that code https://github.com/AvaloniaUI/Avalonia/blob/master/src/iOS/Avalonia.iOS/AvaloniaAppDelegate.cs#L45-L66

**Reply by tkefauver** [reply]
> How do you setup/initialize Avalonia in this project?

Its an ios extension project so there's no `Application.Main` entry point. All I can do is override methods in the `UIInputViewController` :

```
[Register("KeyboardViewController")]
public class KeyboardViewController : UIInputViewController, IAvaloniaViewController {
	// ... IAvaloniaViewController Implementation here
	// ... same as DefaultAvaloniaViewController
	
	UIStackView outerStackView = null;
	AvaloniaView keyboardView = null;
	AppDelegate appDelegate = null;
	
	public override void LoadView() {
		base.LoadView();
		try {
			appDelegate = new AppDelegate();
			appDelegate.FinishedLaunching(UIApplication.SharedApplication, null);
		}
		catch (Exception ex) {
			// log error...
		}
	}
	
	public override void ViewDidLoad() {
		base.ViewDidLoad();
		
		outerStackView = new UIStackView() {
			BackgroundColor = UIColor.Magenta,
			Axis = UILayoutConstraintAxis.Vertical,
			TranslatesAutoresizingMaskIntoConstraints = false,
			Spacing = 5
		};
		View.AddSubview(outerStackView);
		
		try {
			keyboardView = new AvaloniaView() {
				Content = KeyboardMainViewModel.CreateKeyboardView(this)
			};
			outerStackView.AddArrangedSubView(keyboardView);
			NSLayoutConstraint.ActivateConstraints(new NSLayoutConstraint[]{
					keyboardView.HeightAnchor.ConstraintGreaterThanOrEqualTo(500),
					});
		}
		catch (Exception ex) {
			// log error..
		}
	}
}
```
]
`AppDelegate` and `App.axaml` are just barebones:

```
// App.axaml

<Application xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:local="using:iosKeyboardTest"
             x:Class="iosKeyboardTest.iOS.KeyboardExt.App"
             RequestedThemeVariant="Default">
    <Application.Styles>
        <FluentTheme />
    </Application.Styles>
</Application>

//App.axaml.cs

public class App : Application {
    public override void Initialize() {
        AvaloniaXamlLoader.Load(this);
    }
    public override void OnFrameworkInitializationCompleted() {
        if (ApplicationLifetime is not ISingleViewApplicationLifetime singleViewPlatform) {
            return;
        }
		singleViewPlatform.MainView = new MainView {
			DataContext = new MainViewModel()
		};
        base.OnFrameworkInitializationCompleted();
    }
}

// AppDelegate.cs
    [Register("AppDelegate")]
    public partial class AppDelegate : AvaloniaAppDelegate<App>
    {
        protected override AppBuilder CustomizeAppBuilder(AppBuilder builder) {
            return base.CustomizeAppBuilder(builder)
                .WithInterFont()
                .UseReactiveUI()
                ;
        }
    }
```
iOS is very stringent about what you're allowed to do in the `UIInputViewController` overrides. It'll reject the keyboard and say `Connection to plugin was interrupted` and not explain why at all. Things like changing threads or async don't seem to be allowed. 
Since its an extension (.appex not .app) I haven't found a way yet to attach a debugger or get any kind of logging so its hard to tell what's going on... 

I DID have it working on a 17.2 simulator but on my 17.4 device it won't load it and now after fiddling with it more, it doesn't work on the simulator anymore either :/

It seems like it might be some kind of threading access violation issue but I'm not an expert at iOS or Avalonia under-the-hood so just guessing

**Reply by maxkatz6** [reply]
> Its an ios extension project so there's no Application.Main entry point

Yeah, it's not as important. The only thing, you need to build an app somewhere. If you still can use AvaloniaAppDelegate, then it should be fine.

**Reply by maxkatz6** [reply]
> I DID have it working on a 17.2 simulator but on my 17.4 device it won't load it and now after fiddling with it more, it doesn't work on the simulator anymore either :/

Latest simulators are bugged. 
You need to use hacks to make it work:
1. Make sure Metal rendering is enabled (it seems optional, but helped for some). You can do that in CustomizeAppBuilder method with `.With(new iOSPlatformOptions { RenderingMode = [Metal] })` (pseudocode).
2. This hack is required in the csproj:
```
    <PropertyGroup Condition="'$(RuntimeIdentifier)' != 'ios-arm64'">
        <RuntimeIdentifier>iossimulator-x64</RuntimeIdentifier>
    </PropertyGroup>
```

Basically, it doesn't run on iossimulator-arm64. So you need to replace it with iossimulator-x64.

**Reply by tkefauver** [reply]
Thanks for the info. I got it working again now. But its probably pointless there's so many limitations w/ iOS custom keyboards. Like ram usage has to stay < 45mb. I don't think its going to work out

**Reply by tkefauver** [reply]
@maxkatz6 from Instruments the keyboard's memory use  is like ~70mb. I tried both OpenGL and Metal they're basically the same. I know on windows software rendering lowers the footprint. Is there any way to trim it down?

**Reply by maxkatz6** [reply]
You can try to publish app with NativeAOT https://github.com/xamarin/xamarin-macios/blob/main/docs/nativeaot.md

**Reply by tkefauver** [reply]
70mb is running on the device so AOT is already enabled.

**Reply by tkefauver** [reply]
Oh duh I didn't realize aot/nativeAot are different, I'll try that tomorrow!

**Reply by maxkatz6** [reply]
Yeah, it's different. Binary size might be bigger, but it should consume less memory as well - since there is no JIT and in-memory code compilation.

**Comment by maxkatz6** [comment]
You likely can skip IAvaloniaViewController/InitWithController part, as it's only used for insets manager.
